File: session-2/challenge-questions/bfs/minimum_depth_binary_tree.py
# https://leetcode.com/problems/minimum-depth-of-binary-tree/


# This problem follows the Binary Tree Level Order Traversal pattern.
# The difference is that instead of keeping track of all the nodes in a level,
# we will only track the depth of the tree.
# As soon as we find our first leaf node we know the minimum depth of the tree.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


from collections import deque

class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:

        # edge case
        if not root:
            return 0

        # core logic
        container = deque()
        min_depth = 0
        container.append(root)

        while container:
            min_depth += 1

            level_size = len(container)
            for _ in range(level_size):
                curr_node = container.popleft()
                # identify if you've found a leaf node
                if not curr_node.left and not curr_node.right:
                    return min_depth

                # if not add its children
                if curr_node.left:
                    container.append(curr_node.left)

                if curr_node.right:
                    container.append(curr_node.right)


'''

Time complexity
The time complexity is O(N) where 'N' is the total number of nodes in the tree.
* in the worst case we traverse each node once.

Space complexity
The space complexity is O(N)
* in the worst case we are storing every node in a queue

'''

File: session-2/challenge-questions/bfs/minimum_depth_binary_tree.java
// https://leetcode.com/problems/minimum-depth-of-binary-tree/

/*

This solution uses a queue to perform a breadth-first search on the binary tree. 

It starts by adding the root node to the queue, and then enters a loop. 

On each iteration of the loop, it increments the depth by 1 
and examines all of the nodes at the current depth (which are stored in the queue). 

If any of these nodes are leaf nodes (have no children), it returns the current depth. 
Otherwise, it adds the children of the current nodes to the queue and continues the loop.

*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            depth++;
            for (int i = 0; i < size; i++) {
                TreeNode current = queue.poll();
                if (current.left == null && current.right == null) {
                    return depth;
                }
                if (current.left != null) {
                    queue.offer(current.left);
                }
                if (current.right != null) {
                    queue.offer(current.right);
                }
            }
        }
        return depth;
    }
}
